### ISSUE TITLE
Can't communicate over UART.

### ISSUE THREAD

**Issue by overflowz** [firmata/firmata.js#187]
Hey everyone,

Found many likely issues here of this but still was not unable to make it work.

I'm trying to connect Arduino and Raspberry Pi 3 over UART. Tried many ways but all of them failed.

What I have now:

ttyAMA0 port on raspberry -- which works pretty fine. Wrote a custom sketch that echoes back what you sent, tested and worked fine like a charm.

Whenever I'm writing StandardFirmataPlus and using either Johnny-Five or firmata's repl.js, it never gets connected and just hanging.

Any thoughts?

**Comment by rwaldron** [comment]
@soundanalogous any thoughts?

**Comment by soundanalogous** [comment]
Is the baud rate set to 57600 on the Pi 3?

**Comment by overflowz** [comment]
@soundanalogous if I remember correctly -- yes. I started serial port on arduino side on baud rate of 57600 and messages were receiving properly with `echo "test" > /dev/ttyAMA0` with echo sketch.

**Comment by soundanalogous** [comment]
Hummm since you're using StandardFirmataPlus I assume you may be trying to use the Firmata Serial feature to communicate with the Pi 3. If that is the case, then you can't use RX0/TX0 since they are used for the primary communication with the board. You'd need to connect RX1/TX1 or higher. That serial data is then routed through RX0/TX0 to your computer (if you're the Due USB connection).

**Comment by overflowz** [comment]
Hmm, I'll post what I used:
StandardFirmataPlus with:
1) #define FIRMATA_SERIAL_FEATURE
2)
````
Serial1.begin(57600);
Firmata.begin(Serial1);
````
and also used Arduino Due's RX1/TX1 pins rather than RX0/TX0
3) both of them.

USB was completely removed when testing these and still no luck. Am I missing something?

**Comment by soundanalogous** [comment]
#1 and #2 above are for two separate things.

FIRMATA_SERIAL_FEATURE is the ability to use a UART through Firmata. This is different than using a UART as the Transport for Firmata. The Serial Feature is for communicating with things like a GPS receiver that requires a UART connection, or for communicating with another board using a protocol of your own definition.

I'm not sure if you're trying to create a custom Serial protocol to communicate between the Due and the Pi or if you're trying to use Firmata to communicate between the two (where you write a firmata.js script that runs on the Pi to communicate with the Due). If it's the latter, then you don't need StandardFirmataPlus. That still doesn't resolve your issue, but it will help me understand exactly what your setup is and what you're trying to accomplish here.

**Comment by overflowz** [comment]
as for FIRMATA_SERIAL_FEATURE, I was just playing with it -- was not sure what it was for.

As for the setup, I'm literally trying to communicate rpi to arduino due using firmata with UART (RX/TX pins) rather than USB one.

**Comment by soundanalogous** [comment]
TLDR: I think the issue may actually be that there is no auto-reset on the Board when you wire directly to the UART.

Historically Arduino boards reset upon establishing a serial connection. This was done by wiring the DTR pin in the UART connection. This meant any Sketch on the board would restart upon establishing a serial connection. For backwards compatibility when boards started using a USB serial, they kept this convention (I'm not sure how it's wired up exactly to achieve this). Firmata adapted this behavior early on (back when there were only a couple of Arduino boards) by sending the Firmata version number when the Sketch starts. Firmata client libraries then adapted this a start signal. If this signal is missing, the client library won't work (very fragile today, but it was common for many years). This really needs to be re-architected so that the client does not need to respond to a start signal from the board.

To test if this is actually the issue, you can try to run the script on your Pi, then immediately hit the hardware reset on the Due. If that works, then indeed this is the issue. It may take a few tries to get the timing of when to press the HW button just right. You may need to extend the firmata.js timeout as well.

**Comment by overflowz** [comment]
Well, still no luck. Tried resetting as well, many, many times.

Here's the screenshot of debug information if it helps somehow, after entering port number & clicking reset one time.

![image](https://user-images.githubusercontent.com/7758774/38456133-f6a65f5c-3a91-11e8-9ff1-7a9bbbc49de8.png)

No matter how many times I click reset or trying to do synchronously, result is always the same -- it hangs (johnny-five throws error to check if I'm using correct version of firmata).

Also tried to debug, to make sure that data is coming for sure and it does. Firmata reports version as it should, like it did on Serial Monitor, even if I do `cat /dev/ttyAMA0` but for some reason, it does not work with firmata itself.

am I missing something?

**Comment by soundanalogous** [comment]
You're getting the right sequence on startup. The protocol version message followed by the firmware name/version message. For some reason firmata.js is not responding. Can you paste in your full JavaScript code?

**Comment by overflowz** [comment]
I'm using firmata's repl.js file located in `node_modules/firmata` folder to test it out.

One more question is, why I'm getting huge data over Serial and only two bytes from other Serial ports(1, 2, 3)?

**Comment by soundanalogous** [comment]
Have you tried examples/blink.js?

**Comment by overflowz** [comment]
I did, I'm getting same errors.

**Comment by soundanalogous** [comment]
I have a Pi 3 and a Due, I'll have to replicate your setup to get a better understanding of what the issue is. Won't be able to get to that until tomorrow at the earliest.

**Comment by overflowz** [comment]
Well, sure! I'll write down what I'm using right now:

1) sketch: StandardFirmata / StandardFirmataPlus.
Also modified one, replacing following lines (https://github.com/firmata/arduino/blob/master/examples/StandardFirmataPlus/StandardFirmataPlus.ino#L801-L802) with this:
````
Serial3.begin(57600);
Firmata.begin(Serial3);
````

2) Wire configuration:
Arduino RX3/TX3 (or RX/TX in case of original sketch) -> Rpi TX/RX (GPIO 14 and 15)
Arduino GND -> Rpi GND
Arduino 5V -> Rpi 5V (or with usb cable, tried both).

3) Rpi configuration:
I have disabled console & bluetooth use of UARTs. (followed this guide https://www.raspberrypi.org/documentation/configuration/uart.md)

Double checked on this and communication works as it should, at least echo sketch is working pretty well.

Raspberry is connected to my PC via USB to power it up and also either it powers up arduino or connecting with USB on my PC too, tried them both.

If you'll need more information or I'm missing something, let me know. I'm also not an expert of wiring, but friend of mine is helping and he assures me that everything should be working fine.

Best regards.

**Comment by soundanalogous** [comment]
You should be able to use `Serial` rather than Serial3/2/1. I was unsure initially which method of Serial communication you were referring to.

Here's one more thing to try. This will skip the capability query that normally replies with a 500+ byte packet that may overflow your UART and cause issues.

```javascript
var Board = require("../");

Board.requestPort(function(error, port) {
  if (error) {
    console.log(error);
    return;
  }

  var board = new Board(port.comName, {
    skipCapabilities: true,
    pinCount: 66,
    analogPins: [54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65]
  });

  board.on("ready", function() {
    var pin = 13;
    var state = 1;

    board.pinMode(pin, board.MODES.OUTPUT);

    setInterval(function() {
      board.digitalWrite(pin, (state ^= 1));
    }, 500);
  });
});
```

**Comment by overflowz** [comment]
I'm getting `No Acceptable Port Found`

**Comment by soundanalogous** [comment]
Hummm, sounds like port detection is not working. You can remove the outer Board.requestPort wrapper function and replace `port.comName` with a hard coded string.

**Comment by soundanalogous** [comment]
Also `var Board = require("firmata");` rather than `var Board = require("../");`

**Comment by overflowz** [comment]
It just hangs like this:

![image](https://user-images.githubusercontent.com/7758774/38460360-4d7f1c1a-3ac9-11e8-8a75-a49a53998e2a.png)

as for `require('firmata');` no worries on that.

**Comment by soundanalogous** [comment]
And you're using the PL011 rather than mini UART on the Pi right?

**Comment by overflowz** [comment]
Yup. For some reason, the code you gave me, it triggers 'ready' event when clicking reset button, but LED is not toggling and as I see on the DEBUG screen, it is reading and writing data often (which was not happening before).

**Comment by soundanalogous** [comment]
Hummm so it looks like transmission is working from Arduino to Pi (you're seeing the firmware name and version being logged), but not from Pi to Arduino (LED does not blink).

**Comment by soundanalogous** [comment]
Can you log what bytes are being sent from the Pi to the Arduino?

**Comment by overflowz** [comment]
Not sure how to do that, but I'll try.

**Comment by soundanalogous** [comment]
In firmata.js, insert a log statement at line 798, logging `data`

**Comment by soundanalogous** [comment]
Actually that should be correct, I'd actually need logging on the Arduino side. I'll have to send you a modified version of StandardFirmata.

**Comment by overflowz** [comment]
Would appreciate it.

**Comment by soundanalogous** [comment]
Compile and upload this Sketch to your Due: https://gist.github.com/soundanalogous/e10be1c5c0e4031e5aba0769ce5bc1e8

Then run this test script:
```javascript
var Board = require("firmata");

var board = new Board("/dev/ttyAMA0", {
  skipCapabilities: true,
  pinCount: 66,
  analogPins: [54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65]
});

board.on("ready", function() {
  var pin = 13;
  var state = 1;

  board.pinMode(pin, board.MODES.OUTPUT);

  board.on("string", function (message) {
    console.log(message);
  });

  setInterval(function() {
    board.digitalWrite(pin, (state ^= 1));
  }, 500);
});
```

That should hopefully log:
Due ready
DW 2 : 1
DW 2 : 0
DW 2 : 1
DW 2 : 0
etc

**Comment by soundanalogous** [comment]
actually it may be DW 2 : 16 rather than DW 2: 1

**Comment by overflowz** [comment]
Well, to be honest it does not print anything. It just hangs. If I add DEBUG flag to see if there's something going, it's the same, transmission is happening, yet nothing is being logged without it.

**Comment by soundanalogous** [comment]
But you still get:
READY
StandardFirmata.ino-2.5
?

**Comment by soundanalogous** [comment]
Ah, I forgot to add that in the script.

**Comment by soundanalogous** [comment]
Updated script to add that logging as well.

**Comment by overflowz** [comment]
Yup. added by myself, no worries :)

![image](https://user-images.githubusercontent.com/7758774/38460749-19c1ce56-3ad1-11e8-937d-25eea40a0d43.png)

It just hangs here.

**Comment by soundanalogous** [comment]
in firmata.js, add `console.log("reportversion");` just after [line 738](https://github.com/firmata/firmata.js/blob/master/lib/firmata.js#L738).

Then add `console.log("skipCapabilities");` just after [this line](https://github.com/firmata/firmata.js/blob/master/lib/firmata.js#L748).

Also add `console.log(data);` [here(https://github.com/firmata/firmata.js/blob/master/lib/firmata.js#L616)].

**Comment by overflowz** [comment]
output:

````
data <Buffer f9 02 05 f0 79 02 05 53 00 74 00 61 00 6e 00 64>
reportversion
data <Buffer 00 61 00 72 00 64 00 46 00 69 00 72 00 6d 00 61 00 74 00 61 00 2e 00 69 00 6e 00 6f 00 f7 f0 71 44 00 75 00 65 00 20 00 72 00 65 00 61 00 64 00 79 00 ... >
skipCapabilities
READY
StandardFirmata.ino-2.5
````

**Comment by soundanalogous** [comment]
Also add `console.log("toggle");` in the setInterval block in test.js to see if the script is still running.

**Comment by overflowz** [comment]
it does output toggle function every 500ms.

**Comment by overflowz** [comment]
It was auto generated elipsis. I'll give you full output of that

**Comment by soundanalogous** [comment]
And you have these lines in your test script?
```
board.on("string", function (message) {
   console.log(message);
});
```

**Comment by overflowz** [comment]
Woops, I don't had that.

Output including on string event:
````
data f90205f07902055300740061006e0064
reportversion
data 006100720064004600690072006d006100740061002e0069006e006f00f7f071440075006500200072006500610064007900f7
skipCapabilities
READY
StandardFirmata.ino-2.5
Due ready
toggle
toggle
toggle
toggle
toggle
toggle
...
````

**Comment by soundanalogous** [comment]
And do you have to hit Reset on the board to get that or does it also work without reset?

**Comment by overflowz** [comment]
I have to click RESET all the time or it just hangs.

**Comment by soundanalogous** [comment]
I've added another message to the logging Sketch. Copy [this gist again](https://gist.github.com/soundanalogous/e10be1c5c0e4031e5aba0769ce5bc1e8), compile and upload. That should print "ping" every 10 seconds in the JS console if the Sketch is still running on the board.

**Comment by soundanalogous** [comment]
You can remove the "toggle" log as well since we know the JS script isn't hanging.

**Comment by overflowz** [comment]
output when clicking RESET:
````
data f90205f07902055300740061006e0064
reportversion
data 006100720064004600690072006d006100740061002e0069006e006f00f7f071440075006500200072006500610064007900f7
skipCapabilities
READY
StandardFirmata.ino-2.5
Due ready
data f071700069006e00
data 6700f7
ping
data f071700069006e00
data 6700f7
ping
data f071700069006e00
data 6700f7
ping
data f071700069006e00
data 6700f7
ping
data f071700069006e00
data 6700f7
ping
````

output without clicking RESET:
````
data f071700069006e006700f7
data f071700069006e00
data 6700f7
data f071700069006e00
data 6700f7
data f071700069006e00
data 6700f7
data f071700069006e00
data 6700f7
data f071700069006e00
data 6700f7
````

**Comment by soundanalogous** [comment]
So strange. Absolutely no issues in receiving data from the board, but data is never received by the board.

**Comment by overflowz** [comment]
Can it be caused if pin is broken? For example RX of arduino, but I would not able to write sketch there too, right?

My question is, do you have the same setup and does it work for you?

**Comment by soundanalogous** [comment]
I have not tried to replicate the setup yet. I've never attempted to connect an Arduino to a Pi.

**Comment by overflowz** [comment]
Oh well. Got any suggestions what should I try in general? Like buy new boards, try on different ones, etc.

**Comment by soundanalogous** [comment]
Well, it sounds like it works fine over USB, so I suspect there's a weird timing issue when using the UART (as evident by you needing to hit the reset button on the board). Maybe the UART isn't true full duplex or there's a weird delay in it. You could try lower baud rates like 9600 or 19200. You could also try johnny-five which is built on top of firmata.js (when an Arduino is involved). This document may be helpful: https://gist.github.com/ajfisher/8880882.

I'll let you know if I get around to setting this up locally and have any success.

**Comment by soundanalogous** [comment]
If you have a different Arduino than a Due that is also worth testing. Due has been one of the more problematic boards over the years.

**Comment by overflowz** [comment]
I was following that guide too, but it was written for rpi 2 though. My main project is written in johnny-five, but when I was getting firmata errors from there, thought would be great to try to start debugging from firmata.js. 

Alright, thanks! I'll also try on other boards and will let you know if I get something.

Really appreciate your help!

**Comment by soundanalogous** [comment]
There are comments at the bottom for Pi 3, but it sounds like you have already tried that.

**Comment by overflowz** [comment]
Yeah, I did. I tried many ways and was not able it to work tho. I'll test if lowering baud rate will help now and if not then I'm out of ideas on this and will try it on Mega later, not sure when though.

**Comment by overflowz** [comment]
output of baud 9600 is even weirder:
````
data 00
data e000e000e000e000e0e00000e0e000e000e000e0e0
data e000
data 0000
data e0e00000
data e000e00000
data e0e000e000e0e000
data 00e0e000
data e000e000
data 00e0e000
data 0000e000
data e000e000
data 0000e000
data 0000e000
data 0000e000
data e000e0e000
data 0000e000
data e0e0e0e000
data e000e000
data 00e0e000
data e000e000
data 00e00000
data e000e0e000
data 00e0e000
data e0e0e000
data e0e000
data e000e00000e000
data e0e0e0e000
data e0e000e000
data 00
data 0000
data 0000e000
data e0e000e000
data e000e000
data 0000e000
data e000e000
data e0e0
````

no board ready event fired, no reaction on RESET button, nothing.

same for 115200 baud:
````
data de0e86b8f8
data de0e86b8f8
data 87f2380098a8a898
data c8a8a898c8b898e8
data 982898a8b8f8de0d
data b808c8b898a8d8f8
data de0e86b8f8
data de0e86b8f8
````

I also set option "baud" of board options in test.js as well.

**Comment by overflowz** [comment]
Never mind, baud is not board's property. Tired for now, will continue testing later.

Will post if I get success anyhow.

**Comment by overflowz** [comment]
Seems it worked pretty well on Arduino Mega, even no need to click RESET button. Both johnny-five, the test.js you gave and repl.js started working fine. I guess it's problem with the Arduino Due itself.

Please, if you manage to test it out on Due, post any updates here, thank you.

**Comment by soundanalogous** [comment]
Which OS are you running on your Pi 3?

**Comment by overflowz** [comment]
Raspbian -- latest one.

**Comment by soundanalogous** [comment]
Ok, trying to get that up and running.

Have you run into this error when trying to run a J5 script on the Pi 3?

```
+---------------------------------------------------------+
|Sorry, you don't have permission to run this program.    |
|Try running as root, e.g. precede the command with sudo. |
+---------------------------------------------------------+
```

**Comment by overflowz** [comment]
Uhm no. Tried it right now but I don't get that error either. I was working with root from the beginning though (sudo su).

**Comment by soundanalogous** [comment]
sudo su works. Strange that is required.

**Comment by soundanalogous** [comment]
I was able to get it working on the Due without issue. This is the output you should see when it works:

```
READY
StandardFirmataLog.ino-2.5
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
ping
DW 1 : 0
DW 1 : 32
DW 1 : 0
DW 1 : 32
DW 1 : 0
```

I suspect the RX0 pin on your Due may be damaged. Also I'm not sure how you're powering your Due. I have mine connected to my PC via the Programming Port, just to give it power. If you have it connected to the Pi for Power, you could try an external supply or connecting to the PC as I have in case that makes a difference.

**Comment by soundanalogous** [comment]
Also, I should note that I did not have to press the RESET button on the Due.

**Comment by overflowz** [comment]
what I tried was -- powering it up from PC via USB, from Raspberry via USB and just share 5v from raspberry with same ground. I guess that's the case that RX0 might be damaged.

Also, can you try doing same thing but with a different Serials? I had same thoughts that RX0 might be damaged so I switched to RX1/TX1 instead and then I get weird results.

EDIT: Oh well, that's pretty weird. Might that also a problem if this Due is bought since the day it was released?

**Comment by soundanalogous** [comment]
I'll try it with RX1/TX1 this evening. My Due is really old as well. Although I may have upgraded the firmware for the ATMega16u at some point. Try this: https://www.arduino.cc/en/Hacking/Upgrading16U2Due

**Comment by soundanalogous** [comment]
I tried using RX1/TX1 but with no luck unfortunately. There must be something in the timing of the startup sequence that causes issues if auto-reset is not working properly. The reason why RX0/TX0 works is because those pins are wired to the ATMega16u that is also wired to reset, sending data from the Pi to the Arduino must trigger RESET in the same way that the USB is wired. For some reason with your Due (either due to the older ATMega16u firmware or a damaged RX0 pin) the auto-reset isn't working properly and the timing is too finicky to trigger it manually (not that would be a solution even if a manual trigger would work).

It will probably take some time to rework the init sequence in firmata.js to cover this use case in a way that doesn't break backwards compatibility with existing applications.

**Comment by overflowz** [comment]
For now, I'll stick to MEGA. After I'll upgrade firmware or get new one (also with upgrading) -- will report results back here, but not yet sure when it will happen though.

Thank you for your help. I'll be still waiting for further updates.

**Comment by soundanalogous** [comment]
Good news! I was able to get `Serial1` to work. Doesn't require manually resetting the board either. However, it does take a few seconds to connect (a tip on how to avoid that delay is below). Yesterday I forgot that I had to ignore the RX1/TX1 pins or the TX pin would be reconfigured as an INPUT pin when Firmata started up. Change your `setup` function to match this:

```c++
void setup()
{
  Firmata.setFirmwareVersion(FIRMATA_FIRMWARE_MAJOR_VERSION, FIRMATA_FIRMWARE_MINOR_VERSION);

  Firmata.attach(ANALOG_MESSAGE, analogWriteCallback);
  Firmata.attach(DIGITAL_MESSAGE, digitalWriteCallback);
  Firmata.attach(REPORT_ANALOG, reportAnalogCallback);
  Firmata.attach(REPORT_DIGITAL, reportDigitalCallback);
  Firmata.attach(SET_PIN_MODE, setPinModeCallback);
  Firmata.attach(SET_DIGITAL_PIN_VALUE, setPinValueCallback);
  Firmata.attach(START_SYSEX, sysexCallback);
  Firmata.attach(SYSTEM_RESET, systemResetCallback);

  // Ignore TX1 and RX1 pins
  Firmata.setPinMode(18, PIN_MODE_IGNORE);
  Firmata.setPinMode(19, PIN_MODE_IGNORE);

  // to use a port other than Serial, such as Serial1 on an Arduino Leonardo or Mega,
  // Call begin(baud) on the alternate serial port and pass it to Firmata to begin like this:
   Serial1.begin(57600);
   Firmata.begin(Serial1);
  // However do not do this if you are using SERIAL_MESSAGE

//  Firmata.begin(57600);
  while (!Serial1) {
    ; // wait for serial port to connect. Needed for ATmega32u4-based boards and Arduino 101
  }

  systemResetCallback();  // reset to default config
}
```

To avoid the long delay on start, add the reportVersionTimout property to your Board constructor in test.js:

```javascript
var board = new Board("/dev/ttyAMA0", {
  reportVersionTimeout: 1000
});
```

You can try values lower than 1000 as well. The timeout always happens because auto-restart doesn't work with Serial1/2/3 so you can lower the timeout to make reduce the delay.

**Comment by overflowz** [comment]
Confirmed, works here too!!

Thank you so much!!
